cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/ZooKeeperEraser.java

```diff
@@ -0,0 +1,218 @@
+     * @param supplier                  Supplier with Kubernetes Resource Operators
+     * @param kafkaAssembly             The Kafka custom resource
+     */
+    public ZooKeeperEraser(
+            Reconciliation reconciliation,
+            ClusterOperatorConfig config,
+            ResourceOperatorSupplier supplier,
+            Kafka kafkaAssembly
+    ) {
+        this.reconciliation = reconciliation;
+        // NOTE: we are going to use the eraser for deleting ZooKeeper so we are not interested in providing old storage and replicas parameters
+        this.zk = ZookeeperCluster.fromCrd(reconciliation, kafkaAssembly, config.versions(), null, 0, supplier.sharedEnvironmentProvider);
+        this.isNetworkPolicyGeneration = config.isNetworkPolicyGeneration();
+
+        this.strimziPodSetOperator = supplier.strimziPodSetOperator;
+        this.secretOperator = supplier.secretOperations;
+        this.serviceAccountOperator = supplier.serviceAccountOperations;
+        this.serviceOperator = supplier.serviceOperations;
+        this.pvcOperator = supplier.pvcOperations;
+        this.storageClassOperator = supplier.storageClassOperations;
+        this.configMapOperator = supplier.configMapOperations;
+        this.networkPolicyOperator = supplier.networkPolicyOperator;
+        this.podDisruptionBudgetOperator = supplier.podDisruptionBudgetOperator;
+    }
+
+    /**
+     * The main reconciliation method which triggers the whole reconciliation pipeline. This is the method which is
+     * expected to be called from the outside to trigger the reconciliation.
+     *
+     * @return              Future which completes when the reconciliation completes
+     */
+    public Future<Void> reconcile()    {
+        LOGGER.infoCr(reconciliation, "Deleting all the ZooKeeper related resources");
+        return jmxSecret()
+                .compose(i -> networkPolicy())
+                .compose(i -> serviceAccount())
+                .compose(i -> service())
+                .compose(i -> headlessService())
+                .compose(i -> certificateSecret())
+                .compose(i -> loggingAndMetricsConfigMap())
+                .compose(i -> podDisruptionBudget())
+                .compose(i -> podSet())
+                .compose(i -> deletePersistentClaims());
+    }
+
+    /**
+     * Deletes the secret with JMX credentials when JMX is enabled
+     *
+     * @return  Completes when the JMX secret is successfully deleted
+     */
+    protected Future<Void> jmxSecret() {
+        return secretOperator.getAsync(reconciliation.namespace(), zk.jmx().secretName())
+                .compose(currentJmxSecret -> {
+                    if (currentJmxSecret != null) {
+                        return secretOperator.reconcile(reconciliation, reconciliation.namespace(), zk.jmx().secretName(), null)
+                                .map((Void) null);
+                    }
+                    return Future.succeededFuture();
+                });
```

Comment:
Is this the right way? AFAIK you do one GET here and reconcile does another GET and DELETE. You should consider:
* Doing only reconcile
* Doing only deleteAsync

If you expect this to run only once, maybe `deleteAsync` is easiest? Same applies to some of the methods below.

Reply:
Maybe I am overthinking here and checking that the "feature" is enabled before actually doing a deletion by reconciling (the same applies to network policies as well).
For most of the other I am just reconciling but as you said it does a GET and DELETE.
By using just `deleteAsync` would be see some Kubernetes error in the operator if the corresponding resource doesn't exist, or? i.e. network policy generation is not enabled and we are trying to delete them.
Couldn't be users warned by seeing this kind of errors?
Of course, using `deleteAsync` directly would be much easier and straightforward even taking into account that this step runs just one time at the end of a migration.

Reply:
I simplified just by using a `deleteAsync` which anyway through the `internalDelete` runs a pre-check which does a GET AFAICS to avoid deleting it if doesn't exist. But, in general, the overall code is much better.